django/utils/url_config.py

```diff
@@ -0,0 +1,135 @@
+import inspect
+from collections import defaultdict, namedtuple
+from urllib import parse
+
+from django.core.exceptions import ImproperlyConfigured
+from django.utils import module_loading
+
+BACKENDS = defaultdict(dict)
+
+BaseParsedURL = namedtuple('ParsedURL', 'scheme username password hostname port path options')
+
+
+class ParsedURL(BaseParsedURL):
+    @property
+    def netloc(self):
+        result = self.hostname
+        if self.port:
+            result += ':{0}'.format(self.port)
+        return result
+
+
+def parse_url(url):
+    """
+    A method to parse URLs into components that handles quirks
+    with the stdlib urlparse, such as lower-cased hostnames.
+    Also parses querystrings into typed components.
+    """
+    if isinstance(url, ParsedURL):
+        return url
+    elif isinstance(url, parse.ParseResult):
+        parsed = url
+    else:
+        parsed = parse.urlparse(url)
+
+    # parsed.hostname always returns a lower-cased hostname
+    # this isn't correct if hostname is a file path, so use '_hostinfo'
+    # to get the actual host
+    hostname, port = parsed._hostinfo
+
+    query = parse.parse_qs(parsed.query)
+
+    options = {}
+    for key, values in query.items():
+        value = values[-1]
+        if value.isdigit():
+            value = int(value)
+        elif value.lower() == 'true':
+            value = True
+        elif value.lower() == 'false':
+            value = False
+
+        options[key] = value
+
+    path = parsed.path[1:]
+
+    return ParsedURL(
+        parsed.scheme,
+        parsed.username,
+        parsed.password,
+        hostname,
+        str(port),
+        path,
+        options
+    )
+
+
+def _register_backend(backend_type, scheme, path):
+    if scheme in BACKENDS[backend_type]:
+        raise ImproperlyConfigured('Scheme {0} has already been registered as a {1} handler'.format(scheme,
+                                                                                                    backend_type))
+    BACKENDS[backend_type][scheme] = path
+
+
+def _get_backend(backend_type, scheme):
+    if scheme not in BACKENDS[backend_type]:
+        raise ImproperlyConfigured('No {0} handler configured for {1}'.format(backend_type, scheme))
+    return BACKENDS[backend_type][scheme]
+
+
+def configure(backend_type, value):
```

Comment:
Should `configure` be private in the same way/for the same reason that `_register_backend` is?

Reply:
Yep 👍